File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第6章逆序对.py
#lanqiaoOJ 1506
n = int(input())
a = list(map(int,input().split()))
b = sorted(a)
for i in range(n):  a[a.index(b[i])] = i+1
tree = [0] * (n+1)
def lowbit(x):
    return x & -x
def update(x, d):
    while(x < n):
        tree[x] += d
        x += lowbit(x)
def sum(x):
    ans = 0
    while(x > 0):
        ans += tree[x]
        x -= lowbit(x)
    return ans
res = 0
a.insert(0,0)  #在最前面加个0
for i in range(len(a)-1, 0, -1):
    update(a[i], 1)
    res += sum(a[i]-1)

print(res)

 

File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第6章逆序对.cpp
//逆序对 lanqiaoOJ 1506

#include <bits/stdc++.h>
using namespace std;
const int N = 500010;
int tree[N],Rank[N],n;         //注：rank是C++的保留字，这里用Rank
#define lowbit(x)  ((x) & - (x))
void update(int x, int d) {
    while(x <= N) {
        tree[x] += d;
        x += lowbit(x);
    }
}
int sum(int x) {
    int ans = 0;
    while(x > 0){
        ans += tree[x];
        x -= lowbit(x);
    }
    return ans;
}
struct point{ int num,val;} a[N];
bool cmp(point x,point y){
    if(x.val == y.val)   return x.num < y.num;
    return x.val < y.val;
}
int main(){
    scanf("%d",&n);
    for(int i=1;i<=n;i++) {
		scanf("%d",&a[i].val);
		a[i].num = i;         //记录顺序，用于离散化
	}
    sort(a+1,a+1+n,cmp);      //排序
    for(int i=1;i<=n;i++)     //离散化，得到新的数字序列rank[]
		Rank[a[i].num]=i;

	long long ans=0;
	for(int i=n;i>0;--i){     //倒序处理
        update(Rank[i],1);
        ans += sum(Rank[i]-1);
    }
    printf("%lld",ans);
    return 0;
}
